jt_notes.md

```diff
@@ -0,0 +1,127 @@
+
+**For chas:** load_data() - "Place holder function for loading data in emulator". What is the intention of this, and is there a clear idea of what an emulator would look like?
+
+**For me:** this is a general question for Chas: how does he intend the emulator to work? I don't think I've fully followed his emulator stuff in "app". Might need to talk through that one together...
+
+**For chas: IMPORTANT, BUG:** as per my comment in this module, I don't think barrier frames are implemented properly.
+
+**For chas: IMPORTANT, BUG:** I think there is a bug associated with (lack of) numExtraRefFrames in initialise_internal_parameters(), as per my comment there.
+
+**For chas: IMPORTANT, approaching a BUG**: there is the general issue that the code is not doing a great job of precise timing. It determines a delay time before sending the trigger, but then executes a bunch more code. Oh, and that delay time is then treated relative to "current_time", which is set *after* doing the phase-matching. This is going to reduce accuracy and precision, and also makes me even more uncomfortable in terms of future-proofing. I think it would be much better to pass around absolute times, not deltas.
+
+**For chas**: more generally, I wish we were working with frame *objects* that have accompanying metadata (such as timestamp!). That would help solve the above issue about timings in a tidy manner, and also make it a lot easier to generalise to other timebases (such as camera timebase, rather than computer). Oh, and it would also help deal with gaps in the timebase (missing frames), which I think would be a useful thing to cope with, even if right now there might not be any concrete scenario in which frames would be dropped. I fear I may not win that argument about frame objects, since it would be quite a big structural change. But, do you have any alternative ideas about how best to cope with the fact that we may well want to work in terms of timebases other than the computer's own timebase?
+
+Remark **for chas:** currently when we schedule a trigger everything hangs, waiting for the appropriate time to arrive. This stalls camera and sync analysis. Don't you think we should really figure out a way of doing this asynchronously, without blocking (or come to a definite conclusion that blocking is the only reliable enough way to do it on the Pi).
+
+**For me: where do the stages come in?** cli.py sets up the usb/serial link, but where is it used, by whom, and when? Should be clearer after a refactor. Either we need to conditionally interact with the stages, or we do it on a more distributed basis, in response to events. That's more how I do it in my code, and I like that approach.
+
+determine_reference_period.py
+
+**For me and chas:** here and elsewhere, I'm happy to help refactor into a class myself. This module definitely needs some sort of refactoring (see "WTF"/"Ugh" comments).
```

Comment:
Your recommendation for a class here is purely because of the amount of variables being passed around, right?

Reply:
Mostly, yes. And partly to do with setting things up to allow us to substitute other algorithms in future. I prefer to do that substitution by defining a common API and implementing different *subclasses* for different alternative algorithms. I realise that in practice it could probably be done just as well in this case by substituting different modules to import. My preference for classes in this case is probably mostly coming from C++, where there are several really strong reasons why defining an API through a base class is preferable.

Reply:
Okay... so this seems to be what I've just done for OpticalGater? See my other comment above @ determine_reference_period.py:26? I think we might be agreed here but still talking at slightly cross purposes?

Reply:
I think this can probably be resolved, since I think it's all covered by the long, separate discussion about class types and roles. I'll leave it for you to resolve this conversation here if you're happy it's all covered elsewhere now.